contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,292 @@
+      check( producer != _producers.end(), "finalizer " + finalizer_name.to_string() + " is not a registered producer");
+
+      // Basic key and signature format check
+      check(finalizer_key.compare(0, 7, "PUB_BLS") == 0, "finalizer key does not start with PUB_BLS: " + finalizer_key);
+      check(proof_of_possession.compare(0, 7, "SIG_BLS") == 0, "proof of possession signature does not start with SIG_BLS: " + proof_of_possession);
+
+      // Duplication check across all registered keys
+      auto idx = _finalizer_keys.get_index<"byfinkey"_n>();
+      auto hash = get_finalizer_key_hash(finalizer_key);
+      check(idx.find(hash) == idx.end(), "duplicate finalizer key: " + finalizer_key);
+
+      // Convert to binary form. The validity will be checked during conversion.
+      const auto fin_key_g1 = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      const auto pop_g2 = eosio::decode_bls_signature_to_g2(proof_of_possession);
+
+      // Proof of possession check
+      check(eosio::bls_pop_verify(fin_key_g1, pop_g2), "proof of possession check failed");
+
+      // Insert the finalizer key into finalyzer_keys table
+      auto finalizer_key_itr = _finalizer_keys.emplace( finalizer_name, [&]( auto& k ) {
+         k.id                   = _finalizer_keys.available_primary_key();
+         k.finalizer_name       = finalizer_name;
+         k.finalizer_key        = finalizer_key;
+         k.finalizer_key_binary = { fin_key_g1.begin(), fin_key_g1.end() };
+         k.finalizer_key_hash   = get_finalizer_key_hash(finalizer_key);
+      });
+
+      // Update finalizers table
+      auto finalizer = _finalizers.find(finalizer_name.value);
+      if( finalizer == _finalizers.end() ) {
+         // This is the first time the finalizer registering a finalizer key,
+         // mark the key active
+         _finalizers.emplace( finalizer_name, [&]( auto& f ) {
+            f.finalizer_name              = finalizer_name;
+            f.active_finalizer_key_id     = finalizer_key_itr->id;
+            f.active_finalizer_key_binary = finalizer_key_itr->finalizer_key_binary;
+            f.finalizer_key_count         = 1;
+         });
+      } else {
+         // Update finalizer_key_count
+         _finalizers.modify( finalizer, same_payer, [&]( auto& f ) {
+            ++f.finalizer_key_count;
+         });
+      }
+   }
+
+   /*
+    * Action to activate a finalizer key
+    *
+    * @pre `finalizer_name` must be a registered producer
+    * @pre `finalizer_key` must be a registered finalizer key in base64url format
+    * @pre Authority of `finalizer_name`
+    */
+   void system_contract::actfinkey( const name& finalizer_name, const std::string& finalizer_key ) {
+      require_auth( finalizer_name );
+
+      auto producer = _producers.find( finalizer_name.value );
+      check( producer != _producers.end(), "finalizer " + finalizer_name.to_string() + " is not a registered producer");
```

Comment:
This check is for whether the producer is still a registered producer. Would it be possible a registered producer registered some finalizer keys, un-registered itself, and later tried to activate one of its finalizer keys?

Another question, when a producer is un-registered, should its finalizer keys be removed?

Reply:
If the producer unregisters, then it will be excluded the next time `update_elected_producers` gets called. The corresponding finalizer does not need to be immediately removed.

A producer that is unregistered can still change their active finalizer keys. They just won't be added into the finalizer set because they are unregistered.